### ISSUE TITLE
Tensorflow Projector - nearest points in original space

### ISSUE THREAD

**Issue by ironv** [tensorflow/tensorflow#17867]
In [Tensorflow Projector](http://projector.tensorflow.org), the cosine and euclidean distance, and the ranking based on that seems to be incorrect when the number of dimensions for the points is 50 or more.  It is fine when the number of dimensions is 49 or less.  Here's some code to test

```
import numpy as np
from sklearn.metrics.pairwise import cosine_distances, euclidean_distances

npoints,ndims = 4,49
nclose = npoints-1
func = cosine_distances

np.random.seed(seed=123456)
embed = np.random.uniform(size=npoints*ndims).reshape((npoints,ndims))
labels = np.array(['pt%03d'%n for n in range(1,npoints+1)])

np.savetxt('embed-%d-%d.tsv'%(npoints,ndims), embed, delimiter='\t', fmt='%.6f')
np.savetxt('labels-%d.tsv'%(npoints), labels, fmt='%s')

dist = func(embed)
ind = np.argsort(dist)[:,1:nclose+1]

names = labels[ind]
dist = dist[np.arange(labels.shape[0])[:,None],ind]

np.concatenate((labels[:,None],
                np.dstack((names,np.around(dist,4))).reshape((dist.shape[0],-1))), axis=1)
```

The output of the above code is shown below.  For `pt001`, the closest is `pt004` with cosine distance 0.1695, etc.
```
array([['pt001', 'pt004', '0.1695', 'pt003', '0.2716', 'pt002', '0.2897'],
       ['pt002', 'pt003', '0.2345', 'pt004', '0.2365', 'pt001', '0.2897'],
       ['pt003', 'pt004', '0.2099', 'pt002', '0.2345', 'pt001', '0.2716'],
       ['pt004', 'pt001', '0.1695', 'pt003', '0.2099', 'pt002', '0.2365']],
      dtype='<U32')
```
The above code also generates files which can be uploaded to the projector website for confirmation.  The output there is:
```
Nearest points in the original space:
pt004 0.169
pt003 0.272
pt002 0.290
```
Now change ndims to 50 in the above code and re-run.  The output is
```
array([['pt001', 'pt002', '0.1675', 'pt004', '0.1968', 'pt003', '0.2571'],
       ['pt002', 'pt001', '0.1675', 'pt004', '0.2444', 'pt003', '0.246'],
       ['pt003', 'pt004', '0.1943', 'pt002', '0.246', 'pt001', '0.2571'],
       ['pt004', 'pt003', '0.1943', 'pt001', '0.1968', 'pt002', '0.2444']],
      dtype='<U32')
```
Once the new files are loaded, the output from the website is:
```
Nearest points in the original space: 
pt002 1.040
pt004 1.304
pt003 1.538
```
The website does report the number of dimensions correctly at the top (first 49 and then 50).  Also, `Spherize Data` was turned on and off that makes a very slight difference.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by ironv** [comment]
Please note that the issue refers to https://projector.tensorflow.org.

Have I written custom code: __No__
OS Platform and Distribution:  __Chrome on Windows__
TensorFlow installed from: __Not applicable__
TensorFlow version: __Not applicable__
Bazel version: __Not applicable__
CUDA/cuDNN version: __Not applicable__
GPU model and memory: __Not applicable__
Exact command to reproduce: __Included in the initial message__

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jart** [comment]
@dsmilkov Could you take a look at this project issue?

**Comment by dsmilkov** [comment]
Hi @ironv 

It's probably due to the fact that as soon as the data points are >=50 dims, we are doing approximate PCA / T-SNE / NearestNeighbors computations by randomly projecting the data to 50 random directions.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dsmilkov** [comment]
Closing it since for dim>50, we do approximate PCA/T-SNE and show an icon on the UI that this is approximate.

**Comment by ram3ai** [comment]
1. UI is stating that data is projected to 200 dimensions. However, I see approximate calculation for 100 dimensions.
2. So "PCA is approximate" warning is intended to warn about all the results? It's on the opposite side of the page and the wording "Nearest point in the **original space**" is misleading.
Could you please have another look into this?

**Comment by vinayakumarr** [comment]
Is there any option to download the nearest Nneighbors as csv file.

**Comment by FangliangBai** [comment]
> Closing this issue since for dim>50, we do approximate PCA/T-SNE and show an icon on the UI that this is approximate.

Could you please give a link to the source code that calculates euclidean distance?
